### ISSUE TITLE
Marshal.GetDelegateForFunction feature parity request

### ISSUE THREAD

**Issue by migueldeicaza** [dotnet/runtime#13197]
Hello team,

`Marshal.GetDelegateForFunctionPointer` thows an exception when the provided parameter is an `Action<T>`.   Given that the actual function signature is known, this limitation seems unnecessary. 

Mono currently allows this idiom, without throwing an exception, and is very convenient, to avoid producing a bunch of delegates for every combination of parameters during interop.

**Comment by hamarb123** [comment]
This could (not sure how many people rely on this feature of mono) be a more significant problem as people will (presumably) have to / want to transition from the Mono runtime to the core runtime in the future. I think this should be prioritised for .NET 7 / 8 (whenever the transition is required, but also earlier the better).

Luckily for me though, I don't think I'm actively using my workaround anywhere at the moment (my workaround calls `GetDelegateForFunctionPointerInternal` (with every check except the one we're discussing here) with a fallback to the actual method for when it's fixed in the future, hopefully the internal method won't be renamed between now and then).

**Comment by jkotas** [comment]
> my workaround calls GetDelegateForFunctionPointerInternal

Calling internal runtime methods is unsupported. We reserve the right to change the internal runtime methods anytime. Also, this workaround won't behave correctly in all cases.

**Comment by hamarb123** [comment]
> Calling internal runtime methods is unsupported. We reserve the right to change the internal runtime methods anytime.

Yes, I'm aware. I'm just hoping that it won't be renamed before this change hopefully happens. If it does, I'll just have to deal with it.

> Also, this workaround won't behave correctly in all cases.

I'm aware of that too, (I think) I've only ever used it (in the past) for Func & Action with blittable types for all parameters - this seems like something that is likely to work. I'm now using autogenerated code (including non-generic delegates, I did this before code generators) that use function pointers instead.

Also, just another quick question, why is `Marshal.GetDelegateForFunctionPointer(IntPtr ptr, Type t)` (and presumably other methods similar to this) hidden? Is it going to be unsupported in the future?

**Comment by jkotas** [comment]
> why is Marshal.GetDelegateForFunctionPointer(IntPtr ptr, Type t) (and presumably other methods similar to this) hidden?

These methods are not AOT friendly. They are hidden to lead people to the AOT-friendly generic equivalents.

**Comment by tannergooding** [comment]
@jkotas, @jkoritzinsky are we ever actually going to do anything here or is this just going to sit open for a few more years and with more duplicates getting closed against it?

If we do want to do something, we might just try throwing copilot at it and seeing if it can get close enough that someone on the team can finish it out.

Otherwise, it feels prudent to just close it with the reasoning that this is generally a legacy API and that function pointers + UCO are the way to handle things moving forward.

**Comment by jkotas** [comment]
I don't expect us to invest in improving `Marshal.GetDelegateForFunctionPointer`. It is a problematic API for multiple reasons. Our interop best practices recommend UnmanagedCallersOnly instead.

From https://learn.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#general-guidance: DO prefer using function pointers and [UnmanagedCallersOnlyAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute) as opposed to Delegate types, when passing callbacks to unmanaged functions in C#
